internal/controller/scheduler/pod_watcher.go

```diff
@@ -19,111 +20,166 @@ import (
+	ignoreJobs   map[uuid.UUID]struct{}
+
+	// The job cancel checkers query the job state every so often.
+	jobCancelCheckerInterval time.Duration
+
+	// Channels that are closed when a cancel checker should stop.
+	cancelCheckerChsMu sync.Mutex
+	cancelCheckerChs   map[uuid.UUID]*onceChan
+
+	// This is the context passed to RegisterInformer.
+	// It's being stored here (grrrr!) because the k8s ResourceEventHandler
+	// interface doesn't have context args. (Working around an interface in a
+	// library outside of our control is a carve-out from the usual rule.)
+	// The context is needed to ensure job cancel checkers are cleaned up.
+	resourceEventHandlerCtx context.Context
 }
 
-// NewImagePullBackOffWatcher creates an informer that will use the Buildkite
-// GraphQL API to cancel jobs that have pods with containers in the
-// ImagePullBackOff state
-func NewImagePullBackOffWatcher(
-	logger *zap.Logger,
-	k8s kubernetes.Interface,
-	cfg *config.Config,
-) *imagePullBackOffWatcher {
-	return &imagePullBackOffWatcher{
-		logger:      logger,
-		k8s:         k8s,
-		gql:         api.NewClient(cfg.BuildkiteToken, cfg.GraphQLEndpoint),
-		cfg:         cfg,
-		gracePeriod: cfg.ImagePullBackOffGradePeriod,
-		ignoreJobs:  make(map[uuid.UUID]struct{}),
+// NewPodWatcher creates an informer that does various things with pods and
+// Buildkite jobs:
+//   - If a container stays in ImagePullBackOff state for too long, the Buildkite
+//     Agent REST API will be used to fail the job and the pod will be evicted.
+//   - If a container stays in ImagePullBackOff, and the pod somehow got through
+//     all the init containers (including the image pull checks...) the Buildkite
+//     GraphQL API will be used to cancel the job instead.
+//   - If a pod is pending, every so often Buildkite will be checked to see if
+//     the corresponding job has been cancelled so that the pod can be evicted
+//     early.
+func NewPodWatcher(logger *zap.Logger, k8s kubernetes.Interface, cfg *config.Config) *podWatcher {
+	return &podWatcher{
+		logger:                      logger,
+		k8s:                         k8s,
+		gql:                         api.NewClient(cfg.BuildkiteToken, cfg.GraphQLEndpoint),
+		cfg:                         cfg,
+		imagePullBackOffGracePeriod: cfg.ImagePullBackOffGradePeriod,
+		jobCancelCheckerInterval:    cfg.JobCancelCheckerPollInterval,
+		ignoreJobs:                  make(map[uuid.UUID]struct{}),
+		cancelCheckerChs:            make(map[uuid.UUID]*onceChan),
 	}
 }
 
 // Creates a Pods informer and registers the handler on it
-func (w *imagePullBackOffWatcher) RegisterInformer(
-	ctx context.Context,
-	factory informers.SharedInformerFactory,
-) error {
+func (w *podWatcher) RegisterInformer(ctx context.Context, factory informers.SharedInformerFactory) error {
 	informer := factory.Core().V1().Pods().Informer()
 	if _, err := informer.AddEventHandler(w); err != nil {
 		return err
 	}
+	w.resourceEventHandlerCtx = ctx // 😡
```

Comment:
Is `RegisterInformer` only ever called once? Could the value be clobbered here and / or does it matter?

Reply:
`RegisterInformer` should be called once from `controller.Run`, and a quick look through the code shows no other calls, so I'm not concerned.